src/processors/AsyncAPIInputProcessor.ts

```diff
@@ -74,30 +77,80 @@ export class AsyncAPIInputProcessor extends AbstractInputProcessor {
-    // Go over all the message payloads and convert them to models
-    for (const message of doc.allMessages()) {
-      const payload = message.payload();
-      if (payload) {
-        const schema = AsyncAPIInputProcessor.convertToInternalSchema(payload);
-        const newCommonModel =
-          JsonSchemaInputProcessor.convertSchemaToCommonModel(schema, options);
-        if (newCommonModel.$id !== undefined) {
-          if (inputModel.models[newCommonModel.$id] !== undefined) {
-            Logger.warn(
-              `Overwriting existing model with $id ${newCommonModel.$id}, are there two models with the same id present?`,
-              newCommonModel
-            );
-          }
-          const metaModel = convertToMetaModel(newCommonModel);
-          inputModel.models[metaModel.name] = metaModel;
-        } else {
+
+    const addToInputModel = (payload: AsyncAPISchemaInterface) => {
+      const schema = AsyncAPIInputProcessor.convertToInternalSchema(payload);
+      const newCommonModel =
+        JsonSchemaInputProcessor.convertSchemaToCommonModel(schema, options);
+
+      if (newCommonModel.$id !== undefined) {
+        if (inputModel.models[newCommonModel.$id] !== undefined) {
           Logger.warn(
-            'Model did not have $id which is required, ignoring.',
+            `Overwriting existing model with $id ${newCommonModel.$id}, are there two models with the same id present?`,
             newCommonModel
           );
         }
+        const metaModel = convertToMetaModel(newCommonModel);
+        inputModel.models[metaModel.name] = metaModel;
+      } else {
+        Logger.warn(
+          'Model did not have $id which is required, ignoring.',
+          newCommonModel
+        );
+      }
+    };
+
+    // Go over all the message payloads and convert them to models
+    const channels = doc.channels();
+
+    if (channels.length) {
+      for (const channel of doc.channels()) {
+        for (const operation of channel.operations()) {
+          const operationJson = operation.json<OperationObject>();
+
+          if (!operationJson.message) {
+            continue;
+          }
+
+          if ('oneOf' in operationJson.message) {
+            const oneOf: AsyncAPISchemaObject[] = [];
+
+            for (const message of operation.messages()) {
+              const payload = message.payload();
+
+              if (!payload) {
+                continue;
+              }
+
+              oneOf.push(payload.json());
+            }
+
+            const payload = new AsyncAPISchema({
+              $id: channel.id(),
+              pointer: channel.meta().pointer,
+              oneOf
+            });
+
+            addToInputModel(payload);
+          } else {
+            for (const message of operation.messages()) {
+              const payload = message.payload();
+              if (payload) {
+                addToInputModel(payload);
+              }
+            }
+          }
```

Comment:
I am pretty sure that if `operation.messages()` returns more than one, it means that it's within a list of one or more messages i.e. part of `oneOf` and we can add the models to the union type then. That would simplify this a bit 👍 That is if that's the case of course, have you checked @kennethaasan?

Reply:
Good idea. I'll check

Reply:
Yes, that worked :)